engine/resources/exec.go

```diff
@@ -138,9 +141,38 @@ func (obj *ExecRes) Validate() error {
 		}
 	}
 
+	// check that environment variables' format is valid
+	for key := range obj.Env {
+		if !isNameValid(key) {
+			return fmt.Errorf("environment variable name must only contain alphanumeric characters or underscore")
+		}
+	}
+
 	return nil
 }
 
+// TODO: move this function further down in code
```

Comment:
Why have a TODO in here?

Reply:
I thought that this function was out of place in its current position, but I wanted it to be easily visible the first time you were looking at it, since it is related to the neighbouring function.

Reply:
This doesn't matter, I'd see it anyways because `git diff` shows me the changes. But if a future reader of the code (not the patch) would find it easier if they were co-located, then it makes sense to do so. This depends on convention and other things. Eg: do you layout chained functions in standard execution order or dependency order? Or do private util functions go at the end so they don't interleave with the struct methods? (That is what I do.)